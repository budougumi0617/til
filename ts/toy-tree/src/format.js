const chalk = require('chalk');

const displayName = (node) => {
  switch (node.type) {
    case 'file':
      return node.name;
    case 'directory':
      return chalk.cyan(node.name);
  }
};

const formatEach = (nodes, prefix) => {
  let result = '';

  nodes.forEach((node, index) => {
    const edge = index === nodes.length - 1;
    const guide = prefix + (edge ? '`--' : '|--');
    const next = prefix + (edge ? '  ' : '|  ');

    result += `${guide} ${displayName(node)}\n`;

    if (node.type === 'directory') {
      result += formatEach(node.children, next);
    }
  });

  return result;
};

exports.format = (node) => {
  return `${node.name}\n${formatEach(node.children, '')}`;
};