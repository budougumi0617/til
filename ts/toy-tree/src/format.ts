import chalk from "chalk";
import { DirectoryNode, TreeNode } from "./types";

const displayName = (node: TreeNode) => {
  switch (node.type) {
    case "file":
      return node.name;
    case "directory":
      return chalk.cyan(node.name);
  }
};

const formatEach = (nodes: TreeNode[], prefix: string) => {
  let result = "";

  nodes.forEach((node, index) => {
    const edge = index === nodes.length - 1;
    const guide = prefix + (edge ? "`--" : "|--");
    const next = prefix + (edge ? "  " : "|  ");

    result += `${guide} ${displayName(node)}\n`;

    if (node.type === "directory") {
      result += formatEach(node.children, next);
    }
  });

  return result;
};

export const format = (node: DirectoryNode) => {
  return `${node.name}\n${formatEach(node.children, "")}`;
};
