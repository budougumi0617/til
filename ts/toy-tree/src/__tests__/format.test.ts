import { format } from "../format";

describe("format", () => {
  test("only root", () => {
    expect(
      format({
        type: "directory",
        name: "root",
        children: [],
      })
    ).toMatchSnapshot();
  });

  test("nest", () => {
    expect(
      format({
        type: "directory",
        name: "root",
        children: [
          {
            type: "file",
            name: "file1.txt",
          },
          {
            type: "file",
            name: "file2.txt",
          },
          {
            type: "directory",
            name: "dir1",
            children: [
              {
                type: "file",
                name: "deep1.txt",
              },
            ],
          },
          {
            type: "directory",
            name: "dir2",
            children: [
              {
                type: "file",
                name: "deep1.txt",
              },
            ],
          },
        ],
      })
    ).toMatchSnapshot();
  });
});
