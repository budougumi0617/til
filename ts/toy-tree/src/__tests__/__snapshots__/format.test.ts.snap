// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`format nest 1`] = `
"root
|-- file1.txt
|-- file2.txt
|-- dir1
|  \`-- deep1.txt
\`-- dir2
  \`-- deep1.txt
"
`;

exports[`format only root 1`] = `
"root
"
`;
